server/src/main/java/org/apache/druid/client/CachingClusteredClient.java

```diff
@@ -462,6 +462,10 @@ private Set<SegmentServerSelector> computeSegmentsToQuery(
               chunk.getChunkNumber()
           );
           segments.add(new SegmentServerSelector(server, segment));
+          if (server.isEmpty()) {
```

Comment:
Instead of always failing, use a query context flag like `partialResult` with possible values `allow` (default behaviour) and `fail`. We should always log a warning when the result is partial, but fail only when specified.